### ISSUE TITLE
Compiling with Cuda 10?

### ISSUE THREAD

**Issue by MustafaMustafa** [CannyLab/tsne-cuda#30]
Hi,

I am wondering if anyone has tried compiling with CUDA 10. I am getting a bunch of "redefinition" errors", which may or may not be related to trying to compile with cuda 10.

```bash
/home/mustafa/tsne-cuda/src/include/kernels/bounding_box.h:17:21: error: redefinition of ‘volatile int stepd’
 extern __device__ volatile int stepd, bottomd, maxdepthd;
                     ^~~~~
/home/mustafa/tsne-cuda/src/include/kernels/bh_rep_forces.h:17:21: note: ‘volatile int stepd’ previously declared here
 extern __device__ volatile int stepd, bottomd, maxdepthd;
                     ^~~~~
/home/mustafa/tsne-cuda/src/include/kernels/bounding_box.h:17:28: error: redefinition of ‘volatile int bottomd’
 extern __device__ volatile int stepd, bottomd, maxdepthd;
                            ^~~~~~~
/home/mustafa/tsne-cuda/src/include/kernels/bh_rep_forces.h:17:28: note: ‘volatile int bottomd’ previously declared here
 extern __device__ volatile int stepd, bottomd, maxdepthd;
                            ^~~~~~~
/home/mustafa/tsne-cuda/src/include/kernels/bounding_box.h:17:37: error: redefinition of ‘volatile int maxdepthd’
 extern __device__ volatile int stepd, bottomd, maxdepthd;
                                     ^~~~~~~~~
/home/mustafa/tsne-cuda/src/include/kernels/bh_rep_forces.h:17:37: note: ‘volatile int maxdepthd’ previously declared here
 extern __device__ volatile int stepd, bottomd, maxdepthd;
                                     ^~~~~~~~~
/home/mustafa/tsne-cuda/src/include/kernels/bounding_box.h:18:17: error: redefinition of ‘unsigned int blkcntd’
 extern __device__ unsigned int blkcntd;
```

**Comment by DavidMChan** [comment]
This is a bug that has to do with the way that we handled global variables in the original version of the code - basically when we split the headers into multiple files, we inadvertantly created some bugs with the externed variables. If you use the "fit_tsne" branch of the code, it should A) compile with CUDA10 a bit easier B) be significantly more stable and C) be faster. The only issue is that we (as PhD students where this isn't our active area of research) haven't had time to update the python wrappers to support all of the Scikit-Learn features for the new code, so it's a bit trickier to use python with it. We're working on adding all of the support, but it's slow going.

**Comment by MustafaMustafa** [comment]
Thank you for the quick response @DavidMChan. I appreciate the extra work you have to put into this to make it useful to others. 

I seem to be getting the in fit_tsne branch.

```bash
[ 54%] Building NVCC (Device) object CMakeFiles/tsnecuda.dir/src/kernels/tsnecuda_generated_tree_summary.cu.o
/home/mustafa/tsne-cuda/src/include/kernels/bounding_box.h:17:21: error: redefinition of ‘volatile int stepd’
 extern __device__ volatile int stepd, bottomd, maxdepthd;
                     ^~~~~
/home/mustafa/tsne-cuda/src/include/kernels/bh_rep_forces.h:17:21: note: ‘volatile int stepd’ previously declared here
 extern __device__ volatile int stepd, bottomd, maxdepthd;
                     ^~~~~
/home/mustafa/tsne-cuda/src/include/kernels/bounding_box.h:17:28: error: redefinition of ‘volatile int bottomd’
 extern __device__ volatile int stepd, bottomd, maxdepthd;
                            ^~~~~~~
/home/mustafa/tsne-cuda/src/include/kernels/bh_rep_forces.h:17:28: note: ‘volatile int bottomd’ previously declared here
 extern __device__ volatile int stepd, bottomd, maxdepthd;
                            ^~~~~~~
/home/mustafa/tsne-cuda/src/include/kernels/bounding_box.h:17:37: error: redefinition of ‘volatile int maxdepthd’
 extern __device__ volatile int stepd, bottomd, maxdepthd;
                                     ^~~~~~~~~
/home/mustafa/tsne-cuda/src/include/kernels/bh_rep_forces.h:17:37: note: ‘volatile int maxdepthd’ previously declared here
```

Branch:
```bash
~/tsne-cuda$ git branch 
* fit_tsne
  master
```

**Comment by DavidMChan** [comment]
Hmm, that's interesting (since it shouldn't even be compiling those files...). Did you make sure to clear your cmake build folder after switching branches?

**Comment by MustafaMustafa** [comment]
Yes, I deleted the build directory and started from scratch. Let me try again just to be sure.

**Comment by MustafaMustafa** [comment]
Of course, I should have re-done `git submodule update` after switching to fit_tsne branch.

**Comment by MustafaMustafa** [comment]
It is still trying to compile those files actually. To reproduce:

```bash
git clone git@github.com:CannyLab/tsne-cuda.git
 cd tsne-cuda/
git checkout fit_tsne
git branch 
git submodule init
git submodule update
vim CMakeLists.txt # Switched MKL to OFF
cd build/
cmake .. -DBUILD_PYTHON=TRUE -DBUILD_TEST=TRUE -DWITH_ZMQ=FALSE
make -j 10 
```

**Comment by EsamGhaleb** [comment]
I was able to compile tsne-cuda with Cuda 10 using the main branch on Ubuntu 18.04. 
here how my cmake looks like:

`cmake .. -DBUILD_PYTHON=TRUE -DWITH_ZMQ=FALSE WITH_MKL=FALSE `

**Comment by combiz** [comment]
Great work.  If compiling with CUDA10 now works it would be convenient if `conda install tsnecuda cuda10 -c cannylab -c numba` was available.  Thanks.

**Comment by EsamGhaleb** [comment]
Unfortunately, even though I successfully installed tsne-cude with CUDA 10, it is not stable and gives Xid errors such as the following:

> NVRM: Xid (PCI:0000:02:00): 31, Ch 0000000f, intr 10000000. MMU Fault: ENGINE GRAPHICS GPCCLIENT_T1_6 faulted @ 0x1b_01ca9000. Fault is of type FAULT_PTE ACCESS_TYPE_WRITE

**Comment by mariopi27** [comment]
Is there an update on using tsnecuda with CUDA10?

**Comment by rmrao** [comment]
Ok so I just had to compile this with CUDA 10 myself, and I've pushed changes that should remove many compilation errors, etc. to the `fit_tsne` branch. A few steps you may have to do to get his working:

1. Update cmake. It looks like some older versions of cmake won't find CUDA 10 libraries correctly. Instructions for updating cmake can be found [here](https://askubuntu.com/questions/829310/how-to-upgrade-cmake-in-ubuntu).
2. Run the following:
```bash
git clone https://github.com/CannyLab/tsne-cuda
git checkout fit-tsne
git submodule init
git submodule update
cd ./build
cmake ..
make -j32
make
cd ./python
pip install -e .
```

A few notes: I changed defaults so it will compile w/o MKL and with Python automatically, so you no longer need to specify this on the fit-tsne branch. CUDA 10 also fixed the issue with g++ versions > 4.9, so you don't have to specify that either. Finally, the last `make` command should serve to clean up any errors you get due to multi-threaded compilation issues, which seem to occur a lot here.

**Comment by rmrao** [comment]
The changes made above have been propagated to `master` in #39. Both @DavidMChan and I have been able to compile this with CUDA 10.
